File: frontend/de/unika/ipd/grgen/util/XMLDumper.java
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 2.0
 * Copyright (C) 2008 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos
 * licensed under GPL v3 (see LICENSE.txt included in the packaging of this file)
 */

/**
 * XMLDumper.java
 *
 * @author Created by Omnicore CodeGuide
 */

package de.unika.ipd.grgen.util;

import java.io.PrintStream;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Map;

public class XMLDumper {

	private final PrintStream ps;

	private int indent = 0;

	private final String indentString;

	// TODO use or remove it
	// private boolean printingAttributes = false;

	private final Collection<XMLDumpable> visited = new HashSet<XMLDumpable>();

	public XMLDumper(PrintStream ps) {
		this(ps, "  ");
	}

	public XMLDumper(PrintStream ps, String indentString) {
		this.ps = ps;
		this.indentString = indentString;
	}

	public void dump(XMLDumpable dumpable) {
		if(visited.contains(dumpable)) {
			dumpRef(dumpable);
			return;
		}

		visited.add(dumpable);

		Map<String, Object> fields = new HashMap<String, Object>();
		dumpable.addFields(fields);
		String tagName = dumpable.getTagName();

		indent();
		ps.print('<');
		ps.print(tagName);
		ps.print(" id=\"");
		ps.print(dumpable.getXMLId());
		ps.print('\"');

		for(Iterator<String> it = fields.keySet().iterator(); it.hasNext();) {
			Object obj = it.next();
			Object val = fields.get(obj);
			if(!(val instanceof Iterator)) {
				ps.print(' ');
				ps.print(obj);
				ps.print("=\"");
				ps.print(val);
				ps.print('\"');
				it.remove();
			}
		}

		if(!fields.isEmpty()) {
			ps.println('>');
			indent++;
			for(Iterator<String> it = fields.keySet().iterator(); it.hasNext();) {
				Object obj = it.next();

				// the cast was checked some lines above
				Iterator<?> childs = (Iterator<?>) fields.get(obj);
				String tag = obj.toString();

				if(childs.hasNext()) {
					indent();
					ps.print('<');
					ps.print(tag);
					ps.println('>');
					indent++;

					while(childs.hasNext()) {
						Object d = childs.next();

						assert d instanceof XMLDumpable;
						dump((XMLDumpable) d);
					}

					indent--;
					indent();
					ps.print("</");
					ps.print(tag);
					ps.println('>');
				}
			}
			indent--;
			indent();
			ps.print("</");
			ps.print(tagName);
			ps.println('>');
		} else
			ps.println("/>");
	}

	private void dumpRef(XMLDumpable dumpable) {
		indent();
		ps.print('<');
		ps.print(dumpable.getRefTagName());
		ps.print(" id=\"");
		ps.print(dumpable.getXMLId());
		ps.println("\"/>");
	}

	private void indent() {
		for(int i = 0; i < indent; i++)
			ps.print(indentString);
	}

}

File: frontend/de/unika/ipd/grgen/util/XMLDumper.cs
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 8.1
 * Copyright (C) 2003-2026 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3, some components/parts use different licenses (see LICENSE.txt included in the packaging of this file)
 * www.grgen.de / www.grgen.net
 */

/// <summary>
/// XMLDumper.java
/// 
/// @author Created by Omnicore CodeGuide
/// </summary>

namespace de.unika.ipd.grgen.util
{

	using System.Collections.Generic;
	using System.Diagnostics;

	public class XMLDumper
	{
		private readonly PrintStream ps;

		private int indentationLevel = 0;

		private readonly string indentString;

		private readonly ICollection<XMLDumpable> visited = new HashSet<XMLDumpable>();

		public XMLDumper(PrintStream ps)
			: this(ps, "  ")
		{
		}

		public XMLDumper(PrintStream ps, string indentString)
		{
			this.ps = ps;
			this.indentString = indentString;
		}

		public virtual void Dump(XMLDumpable dumpable)
		{
			if(visited.Contains(dumpable))
			{
				DumpRef(dumpable);
				return;
			}

			visited.Add(dumpable);

			IDictionary<string, object> fields = new Dictionary<string, object>();
			dumpable.AddFields(fields);
			string tagName = dumpable.TagName;

			Indent();
			ps.Print('<');
			ps.Print(tagName);
			ps.Print(" id=\"");
			ps.Print(dumpable.XMLId);
			ps.Print('\"');

			IList<string> keysToRemove = new List<string>();

			foreach(string obj in fields.Keys)
			{
				object val = fields[obj];
	// JAVA TO C# CONVERTER WARNING: Java wildcard generics have no direct equivalent in C#:
	// ORIGINAL LINE: if(!(val instanceof java.util.Iterator<?>))
				if(!(val is IEnumerator<object>))
				{
					ps.Print(' ');
					ps.Print(obj);
					ps.Print("=\"");
					ps.Print(val);
					ps.Print('\"');
					keysToRemove.Add(obj);
				}
			}

			foreach(string keyToRemove in keysToRemove)
				fields.Remove(keyToRemove);

			if(fields.Count > 0)
			{
				ps.Println('>');
				indentationLevel++;
				foreach(object obj in fields.Keys)
				{
					// the cast was checked some lines above
	// JAVA TO C# CONVERTER WARNING: Java wildcard generics have no direct equivalent in C#:
	// ORIGINAL LINE: java.util.Iterator<?> childs = (java.util.Iterator<?>)fields.get(obj);
					IEnumerator<object> childs = (IEnumerator<object>)fields[obj];
					string tag = obj.ToString();

					bool hasNext = childs.MoveNext();
					if(hasNext)
					{
						Indent();
						ps.Print('<');
						ps.Print(tag);
						ps.Println('>');
						indentationLevel++;

						do
						{
							object d = childs.Current;

							Debug.Assert(d is XMLDumpable);
							Dump((XMLDumpable)d);
						}
						while(childs.MoveNext());

						indentationLevel--;
						Indent();
						ps.Print("</");
						ps.Print(tag);
						ps.Println('>');
					}
				}
				indentationLevel--;
				Indent();
				ps.Print("</");
				ps.Print(tagName);
				ps.Println('>');
			}
			else
				ps.Println("/>");
		}

		private void DumpRef(XMLDumpable dumpable)
		{
			Indent();
			ps.Print('<');
			ps.Print(dumpable.RefTagName);
			ps.Print(" id=\"");
			ps.Print(dumpable.XMLId);
			ps.Println("\"/>");
		}

		private void Indent()
		{
			for(int i = 0; i < indentationLevel; i++)
				ps.Print(indentString);
		}
	}

}
